Add PolyFieldBase abstract class

chrisphilip322 opened the pull request:
**NOTE** This change supports `marshmallow~=2.0` and `marshmallow-polyfield~=3.0` and I originally branched off of 35cf134da9042b8c84cd693653b126dc23cce778

I didn't like the interface of having to construct instances of PolyField by passing functions to the constructor. This change adds a new base class PolyFieldBase which PolyField is a subclass of. Users can use the old method of passing arguments to the constructor of PolyField or by subclassing PolyFieldBase which is an ABC with abstract methods of serialization_schema_selector and deserialization_schema_selector.

I also updated the unit tests to test the old way and the new way of constructing PolyField instances.

Commit:
allow subclassing of PolyFieldBase to create new PolyField instances

Commit:
bump setup.py

Commit:
add __all__ to remove noqa comment

chrisphilip322 committed:
Merge pull request #1 from macbeth322/feature-subclassable-polyfield

Add Feature to subclass PolyFieldBase

Bachmann1234 commented:
It's gonna take me some time to sit down and review this in detail. Been a while since I've even looked at this project.

But I'll try and get to it this week. Thank you for the pr!

Jamim commented:
The related issue: #11

Bachmann1234 commented:
Ok, so I looked over this. I love the approach and is probably what I should have done in the first place :-P

So we put this on a separate branch. *If you resolve the conflicts I can merge this*. 

It would be nice to bring this approach into the master branch which is currently for marshmallow 3. If you were willing to do that we could put it on master and release it.

But I understand your usecase is 2x. So maybe i'll try and find some time to make a 3.0 version of this approach.

chrisphilip322 committed:
Merge branch 'marshmallow-3.X' of github.com:Bachmann1234/marshmallow-polyfield into release-3-x

Conflicts:
	requirements.txt
	setup.py
	tests/test_deserialization.py

chrisphilip322 committed:
fix merge

chrisphilip322 commented:
I fixed the merge conflicts, but I just moved back to what I originall had. https://github.com/Bachmann1234/marshmallow-polyfield/tree/marshmallow-3.X looks like it is for v4 of `marshmallow-polyfield` but my change was for v3. I will try and apply my change to head of master soon when I get some time.

Sorry for the late reply to this, I completely missed these messages.

Bachmann1234 commented:
No pressure! No expectations. Just whenever you have the time and the desire :-)

chrisphilip322 commented:
I made #14 to add these changes to master branch.

You may want to have a look at the travis configuration, it doesn't seem to be working for python 3.4.

chrisphilip322 committed:
fix coverage

chrisphilip322 committed:
actually fix coverag

chrisphilip322 committed:
update version support

chrisphilip322 committed:
fix version number

chrisphilip322 committed:
update travis yml

Jamim commented on tox.ini:
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist=py27,py33,py34,pypy
+envlist=py27,py36,pypy
```
Hi @macbeth322,
You probably missed `py37` here.

Bachmann1234 replied:
No... 37 requires annoying changes to the Travis config I have not gotten around to yet

Jamim replied:
I suppose you can just add `dist: xenial` into `.travis.yml`.

Jamim replied:
By the way, how is `tox.ini` relevant to `Travis`?

Bachmann1234 replied:
Honestly it may not be at this point. I dont remember he history of why I had it in there. Its possible I used it at one point

Bachmann1234 replied:
Yeah the config I need to do is similar to what I set up here https://github.com/Bachmann1234/Thrift-Explorer/blob/master/.travis.yml I just have not gotten around to it

Bachmann1234 replied:
Honestly reading it now its less annoying then I remember, ill get to it soon. Less relevant to this PR. If it breaks 3.7 ill deal with it

Jamim replied:
I just created a PR to add Python 3.7 for the corresponding branch:
macbeth322/marshmallow-polyfield#2

chrisphilip322 replied:
I've never used travis personally, but if @Bachmann1234 thinks macbeth322#2 looks good I can merge it.

Bachmann1234 replied:
Looks good to me

Jamim commented on .travis.yml:
```diff
@@ -3,8 +3,7 @@
 python:
-  - "3.4"
-  - "3.3"
+  - "3.6"
```
I believe  `3.7` should be there as well.

Bachmann1234 commented:
Im getting home late tonight but if I have the energy ill get this reviewed and likely merged/released. Otherwise ill get it done over the weekend

Bachmann1234 commented:
@Jamim mind opening a PR against  my repo with your python3.7 change? I could do it by hand but I wanna make sure you get the credit

chrisphilip322 commented:
@Bachmann1234 are you planning on releasing a new 3.x version to support marshmallow 2?
